Added es build, which will get beneficial for es2015 modules aware bu…

Andarist opened the pull request:
…ndlers like webpack 2+ and rollup

IMHO this should also be cherry-picked to the `v1` branch.

This will allow webpack3 to use scope hoisting on this package + ofc tree-shaking will become possible.

jquense commented:
Thanks for this! There is a bit of an issue as is. The `/lib` is actually published as the "root" directory of the package, which is why you can do `import TransitionGroup from 'react-transition-group/TransitionGroup'` without including the `lib` bit, essentially if you want `es` directory published in needs to be _in_ the lib one

Andarist commented:
No worries, I've got u covered :) ! 

Ive just ammended the PR, please take a look. Such structure, although slightly inconvenient (`package.json` per each additional 'cherry-picked' module), allows correct filed to be required by each bundler.

jquense commented:
Is there a reason we can't copy the `/es` directory into `l/ib` instead? I really don't want to have to maintain a package.json for each file I export

Andarist commented:
we cannot support `import TransitionGroup from 'react-transition-group/TransitionGroup'` form for `es` consumers if we copy `es` to the `lib`

also i aint sure what `package.json` is geting published, because it aint the one in the git repository and we need to put `module` property in it, so it can be interpreted correctly and point to the correct file

jquense commented:
> we cannot support import TransitionGroup from 'react-transition-group/TransitionGroup' form for es consumers if we copy es to the lib

I'm less concerned about that I think, its still the more uncommon case. Eventually we can flip it. The package.json in the root is the one being published, tho the `main` field is adjusted to `index.js` before it's copied there

taion commented:
Interesting... I wonder if a separate `-es` package like Lodash does might make sense here, if you want to keep top-level exports.

Andarist commented:
> Interesting... I wonder if a separate -es package like Lodash does might make sense here, if you want to keep top-level exports.

I think this only bring more confusion to the table, especially for newcomers. A single package works out of the box for both commonjs and es consumers - it also enables easier migration paths from i.e. webpack1 to the webpack2+. Personally I wouldnt vote for this one.

Personally puting `es` directory in `lib` for publication purposes seems a bit off too to me, but I'll adjust the PR however you want. However even if thats more uncommon for people to import things like this, its still what you allow and what you also expose in the README as viable solution - so making it incorrect for es consumers will take people off guard here. For me personally it wont be that much of a matter, as well - Im here to discuss it and whatever we decide Ill know how to correctly use the library in my app. Others? Not so much.

Let me know what you decide that should be done and I'll adjust the PR.

Andarist commented:
@taion @jquense any decisions about this?

jquense commented:
I'm fine adding this, but we need to add it the lib directory per my comment above :)

Andarist commented:
Any comments regarding concerns raised [here](https://github.com/reactjs/react-transition-group/pull/77#issuecomment-309880734)?

jquense commented:
I think their reasonable concerns! I don't think think having folks cherrypick from an `es` namespace is to much trouble compared to the maintenance burden, and general weirdness of having a package.json and sub directory for each file in the repo.

Andarist commented:
Ok, I will rebase and adjust the PR soon.

Andarist commented:
@jquense Rebased and it should work. However your `release-script` is replacing the `altPkgRootFolder` [only in `main` entry](https://github.com/AlexKVal/release-script/blob/ab6e0a090cfa830a4ab2f14b651e1175a077725e/src/release.js#L392). So in the real `package.json` main points now to pre-release-script path, while module points to the post-release-script path.

I don't think this release-script was designed with multiple entry points in mind and personally I think you'd be better without it, but that's your choice.

Let me know if everything looks ok :)

Andarist commented:
@jquense friendly 🏓 :)

Andarist commented:
@silvenon I see you are active lately here, so maybe u'd find some time to comment on this one? 😉

Andarist commented on package.json:
```diff
@@ -3,11 +3,14 @@
+    "build": "rimraf lib && npm run build:cjs && npm run build:es && npm run build:dist && cp README.md LICENSE ./lib",
+    "build:es": "cross-env BABEL_ENV=es babel src --out-dir lib/es",
+    "build:cjs": "cross-env BABEL_ENV=cjs babel src --out-dir lib",
     "build:dist": "rimraf lib/dist && webpack && NODE_ENV=production webpack -p",
```
sure thin, gonna apply this change when we agree on other stuff (would like to make a single, bigger change than multiple small ones)

Andarist commented on package.json:
```diff
@@ -3,11 +3,14 @@
   "version": "2.3.0-beta.0",
   "description": "A react component toolset for managing animations",
   "main": "lib/index.js",
+  "module": "es/index.js",
```
Yes, although IMHO publishing `lib` is obscure and personally I would advocate for removing this and just publish normally pkg root + filter files with `"files"` entry in `package.json`

Andarist commented on .babelrc:
```diff
@@ -1,15 +1,5 @@
-    "add-module-exports",
-    "transform-object-assign",
-    ["transform-react-remove-prop-types", { "mode": "wrap" }]
+    "./.babelrc.js"
```
😉 I never have trusted babel's env either, I'm rly glad that I have found this neat trick 😀

silvenon commented:
Thanks for explaining it so well! These additional `package.json` files don't seem so bad, also we can auto-generate them, as you said. I would like to go that route to keep backwards incompatibility. Also, I don't see a problem with having `lib` and `es` side by side if nothing would change for users.

I'll take some time to study your blog post and see how we can make all this work with our release process. Then we'll present everything to the owners to get their OK.

silvenon commented:
@jquense @taion if we add ES modules without changing how the user consumes the package, so that if they turn ES modules on they just get tree-shaking for free, would that be ok if there's little-to-no maintaining overhead on our part? Is there another reason why you're tied to the current `lib`-as-root solution?

Andarist commented:
Lib as root approach gives more natural “deep” imports, otherwise users would have to use react-transition-group/lib/Transition. But as I said the same thing can be achieved with additional package.json files (just pointing out the possible reason behind lib as root approach)

Andarist commented:
And ofc - this shouldnt be a breaking change for the users in any way :)

jquense commented:
I mentioned above that managing a bunch of additional package.json files is not something i really want to do, even tho they can be generated, less is more when it comes to build scripts.  If someone wants to do the work there and can produce some reusable, small thing we can maybe stick in it's own package i'd be open to it, since it is a nicer experience, but otherwise i want to side on not having a lot of bespoke builds since i work on a lot of stuff

We do publish lib/ as the root for cleaner imports, since the _recommended_ approach, even with esm builds, is cherry-picking, since tree-shaking is an imperfect optimization

Andarist commented:
> If someone wants to do the work there and can produce some reusable, small thing we can maybe stick in it's own package i'd be open to it, since it is a nicer experience, but otherwise i want to side on not having a lot of bespoke builds since i work on a lot of stuff

Sure, can work on that 😉

> We do publish lib/ as the root for cleaner imports, since the recommended approach, even with esm builds, is cherry-picking, since tree-shaking is an imperfect optimization

With `cherry-pick`ing tool we'll be able to ditch `lib` publishing and provide cleaner imports generated by a tool 👍

Andarist commented:
@silvenon @jquense 

I've created [`cherry-pick`](https://github.com/Andarist/cherry-pick) to create those additional proxy directories automatically. It has also a separate `clean` method which can be used as postpublish hook to keep the package directory clean.

I've tweaked the PR to:
- use my `cherry-pick` tool
- removed `release-script` and corresponding scripts, because you are using `semantic-release` now and those seemed like some leftovers after the transition
- done few other tweaks to accomodate for new setup

jquense commented:
Hey @Andarist  thanks for all the work here, That tool is neat! I was chatting a bit about this with @taion yesterday and he noted that it may be enough to just export `.mjs` files along side the cjs ones. Modern webpack seems to do the "right" thing in that case, where it will resolve to the `mjs` file in the appropriate environment

Andarist commented:
That's true - I wouldn't like to bet on `.mjs` just yet though, as we know there are some strong feelings in the community against this.

Also there are downsides of that solution:
- for the time being webpack is the only major tool that I know of that supports it out of the box
- works only in webpack@4 (which is still a fresh and not adopted by many people) without additional configuration

To support it you'd have to specify `"main"` in package.json as `"main": "lib/index"` or `"main": "lib"` (basically you have to omit the extension of the file, optionally you can really on `index` being auto-detected as default file for requested directory).

taion commented:
@Andarist 

`package.main` doesn't matter, as `module` would still apply when using the top-level export. This would just be for nested imports without extensions, e.g. `foo/lib/Bar` would resolve to `foo/lib/Bar.mjs` when available, which is what's desired.

This is the default behavior in webpack 4, and it looks like there's a pending PR for Rollup https://github.com/rollup/rollup-plugin-node-resolve/pull/151/files.

I don't think this is really much of a "bet" per se; this should "just work" for any users with a bundler that supports this pattern. For users on a bundler that doesn't support this pattern, they'll just fall back to CJS, but everything should still work.

Andarist commented:
> I'm not sure using the file extension amounts to "betting" on much, in a webpack world we are already consigning ourselves to work the way webpack works. [...] I agree that it's all still a bit up in the air, but its also completely transparent to a consumer so i don't really mind that so much.

I somewhat agree with that. My main concern is that this solution targets webpack@4 or enforce custom configs for previous version (and this suffers from discoverability issue). There are still plenty of apps using webpack@3, webpack@2 and even webpack@1 (although I feel sorry for those users :wink:) out there. Going with proxy directories would allow both previous versions of webpack to "just work" with it without users even noticing or changing anything in their configs.

> That doesn't seem like a problem, app bundling market share is like 99.9% webpack, and rollup is the only sort-of alternative, but that's solely for libraries, which is really relevant here.

With recent rollup advancements that's a statement a little bit underestimating it. It now supports code splitting, dynamic imports and other things - it's starting to becoming a real alternative for smaller projects which don't have to process other assets (like css, images, etc) with it. 

> It should be fine to to do main: "lib/index.js" along withmodule: "lib/index.mjs"` for that, same as we have to do now.

Yeah, considering only web bundlers that's true :+1:

> Sorry, i'm not trying to be difficult here

Not a problem, I enjoy discussing this stuff :wink:

> to me it seems like both approaches work at the whim of bundler defaults, but the directory package.json approach involves a lot more file resolution, i'd be concerned as a general approach that maybe it's less performant?

Hard to tell without actually measuring this, I wouldn't suspect this be that much slower - resolving algorithms has to check bunch of stuff anyway. `.mjs` might be a little bit faster as it's one of the first things to be checked. IMHO it's not much of an issue anyway as this is build time cost only (and probably negligible one), I care much more about runtime costs in the browsers.

> I don't think this is really much of a "bet" per se; this should "just work" for any users with a bundler that supports this pattern. For users on a bundler that doesn't support this pattern, they'll just fall back to CJS, but everything should still work.

That's also true. I'd only like to provide best possible, out of the box, experience for wider range of users.

Either way - if you feel that `.mjs` is better solution considering all your and mine arguments then I'm fine with it. Let me know and I'll adjust this PR accordingly.

Andarist commented:
Any consensus on this? 😉

taion commented:
I think we should just do the `.mjs` thing. I think @jquense was working on pulling some of the tooling used in graphql-js out into a library that could just be used as a dev dep.

jquense commented:
Let's give mjs a shot first since it's the lowest cost change. This is all a bit experimental so let's see how that goes and adjust if necessary

Andarist commented:
I've tweaked the PR to use .mjs strategy. I think I thought about everything and this should just work - but because of unusual (to me) setup with semantic-release & publishing alternate directory I would prefer releasing `next`/`beta` version first, just to check if everything is OK & then release is as `latest`.

Andarist commented:
Damn, tests are failing. Unfortunately I have to go now, gonna check it 2morrow. Sorry for the false alarm 😉

silvenon commented:
You can fix these tests by appending `.default` to these two require calls:

https://github.com/reactjs/react-transition-group/blob/84af1ee730097847eb23eaeeb52acb0535a8b763/test/TransitionGroup-test.js#L17

https://github.com/reactjs/react-transition-group/blob/84af1ee730097847eb23eaeeb52acb0535a8b763/test/CSSTransitionGroup-test.js#L28

Andarist commented:
> You can fix tests by appending .default to these two require calls:

That would actually indicate a breaking change. I've fixed the problem - babel plugin ordering was the problem.

This should be ready now for the final review 😉

Andarist commented:
@silvenon what about making a beta release with this?

jquense commented:
That's a good idea. Sorry for the delay here, we're all a bit busy!

Andarist commented:
Not a problem! I totally understand it, just pinging from time to time because I know how easy it is to miss PRs/issues after time. Take your time 😃

silvenon commented:
Do you think you could also fix #147?

I don't have much experience with releasing betas, for that to work we would need to create a separate branch, right? Would you then be able to redirect this PR to that branch? Sorry for newbish questions. 😅

Andarist commented:
> Do you think you could also fix #147? That would be a breaking change, but seems like it would fit well here.

It won't fix it because `Transition` exports more than a default export and thus this cannot be fixed `babel-plugin-add-module-exports`. It **might** be fixable with other custom setup for this CJS entry, but that would have to be fixed with another PR (and figured out first 😉) - not sure if it's worth the hassle though.

> I don't have much experience with releasing betas, for that to work we would need to create a separate branch, right?

I'm also a little bit unsure how to do that with current semantic release setup. I guess it can be a manual process - create a branch, merge to it, do a beta release from ur machine. But maybe it's configured somehow to release from some beta branch? Dont know.

> Would you then be able to redirect this PR to that branch? Sorry for newbish questions. 😅

Sure thing.

silvenon commented:
I created a `v3` branch. Let's merge that in there and we'll figure out what to do next later. Let me know if you need help with resolving conflicts. (I believe Yarn has some kind of an auto-resolve feature.)

Andarist commented:
I've rebased it and changed the base of this PR to `v3`. Im not sure if u want actually release this as new major though - those changes should be 100% compatible with v2 line

silvenon commented:
Thanks for resolving conflicts! If this change is backwards compatible I definitely won’t bump the major version. I’ll merge within a day or two, I promise, I just want to be careful and study our release process so I don’t screw anything up.

jquense commented:
@silvenon you can merge into any branch apart from `master` without fear of semantic-release, it will only release from master. I don't have a good sense of how to use SR to do prereleases so I think we'll probably just have to do that manually, but it may be worth asking some folks that are better with SR

silvenon commented:
Ok, I created a new `dev` branch, which is same as `v3`, but has a better name. There we can manually release what we want. I wish there was a way to just rename the branch on GitHub, but I'll have to ask you to direct your branch to `dev`, then I'll merge.

Thank you very much! ❤️

Andarist commented:
@silvenon what's the status of this? Is there anything I could do to help moving this forward?

silvenon commented:
@Andarist thanks for pinging, I need to resolve conflicts if there are any and check for myself if this works correctly, maybe add a test if possible. I guess I'm still insecure about my knowledge in types of JavaScript modules so I don't want to merge something breaking, even though I trust that you know and it's really not a lot of code.

It would help if you could tell me how to test manually whether this works. Is it enough to set `modules: false` in `@babel/preset-env` and newest webpack will automatically pick up the ES build?

TrySound commented:
mjs way will prevent loading esm from module field in react-lifecycles-compat. Webpack is very opinionated about this.

Andarist commented:
@TrySound could u elaborate? Preferably with example

TrySound commented:
From what I see package with module field (like material ui in the future) will prefer main field over mjs. Package with mjs (like this one) will prefer main field over module field (in case of react-lifecycles-compat). This makes mjs quite unusable since in most cases you get just cjs.
Webpack sucks.

TrySound commented:
I personally don't care a lot because my production bundler is rollup.

jquense commented:
Wepback will resolve any mjs file over here given how it's extensions are set for defaults. What it won't do tho is interop with  cjs pretending to be esm tho, it's mjs parsing is strict which means any import that's not mjs is considered to be cjs and can only have default imports. This conflicts with webpacks other resolution logic which prefers the module entry to main. It's all sort of a mess for interop, but webpack is trying to do no harm by implementing mjs logic not to spec or out of line with what node is doing, to help prevent longer term fragmentation

jquense commented:
To be slightly cleaerer mjs files in webpack will only consider other miss files to be esm modules, not esm in .js files

Andarist commented:
O damn, this is is a mess then - packages have and will have all sort of dependencies and it's IMHO impractical to require each dependency to use `.mjs` to be treated as ESM module.

I would personally be in favour of not using `.mjs` in this situation and just leverage more "standard" (as in - not conflicting with anything) "nested" package.json files, but this was met with hesitance previously.

jquense commented:
Yeah sorry @Andarist i realize that I did all this exploration into how this works with other of my own packages and didn't follow back up here. 

Let's move to the nested package.json route and get something out? Care to take that on?

Andarist commented:
> Let's move to the nested package.json route and get something out? Care to take that on?

Yeah, sure! I think @TrySound might have wanted to do that too, could you confirm Bogdan? If you are not going to work on this one I'll probably get to it in few days.

TrySound commented:
Sure, I can do it too.
